File: 144-binary-tree-preorder-traversal/144-binary-tree-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> list = new ArrayList<>();
        return pre(root, list);
    }
    
    private List<Integer> pre(TreeNode root, List<Integer> list) {
        if (root == null)
            return list;
        list.add(root.val);
        pre(root.left, list);
        return pre(root.right, list);
    }
}

File: 144-binary-tree-preorder-traversal/144-binary-tree-preorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> ans;
        return pre(root, ans);
    }
    
    vector<int> pre(TreeNode* root, vector<int>& ans) {
        if (root == NULL)
            return ans;
        ans.push_back(root->val);
        pre(root->left, ans);
        return pre(root->right, ans);
    }
};